pkg/segment/structs/segstructs.go

```diff
@@ -961,6 +973,201 @@ func (qtype QueryType) String() string {
 	}
 }
 
+func (qa *QueryAggregators) HasStatsBlock() bool {
+	return qa != nil && !qa.HasStreamStats() && qa.HasGroupByOrMeasureAggsInBlock()
+}
+
+func (qa *QueryAggregators) HasStatsBlockInChain() bool {
+	if qa == nil {
+		return false
+	}
+
+	if qa.HasStatsBlock() {
+		return true
+	}
+
+	if qa.Next != nil {
+		return qa.Next.HasStatsBlockInChain()
+	}
+
+	return false
+}
+
+// returns all columns in the query aggregators if stats are present.
+// Update this function whenever a new struct or an is added to the query aggregators.
+func (qa *QueryAggregators) GetAllColsInAggsIfStatsPresent() map[string]struct{} {
+	if qa == nil {
+		return nil
+	}
+
+	if !qa.HasStatsBlockInChain() {
+		return nil
+	}
+
+	cols := make(map[string]struct{})
+
+	qa.GetAllColsInAggsChainUntilStatsBlock(cols)
+
+	return cols
+}
+
+func (qa *QueryAggregators) GetAllColsInAggsChainUntilStatsBlock(cols map[string]struct{}) {
+	if qa == nil {
+		return
+	}
+
+	AddAllColumnsInOutputTransforms(cols, qa.OutputTransforms)
+	AddAllColumnsInMeasureAggs(cols, qa.MeasureOperations)
+	AddAllColumnsInGroupByRequest(cols, qa.GroupByRequest)
+	AddAllColumnsInTransactionArguments(cols, qa.TransactionArguments)
+	AddAllColumnsInStreamStatsOptions(cols, qa.StreamStatsOptions)
+
+	if qa.HasStatsBlock() {
+		// We want to stop at the first stats block
+		return
+	}
```

Comment:
This looks like it gets all the columns through the first stats block rather than stopping just before the first stats block. Is that right? If so, I think you should change the name to indicate that

Reply:
Yes we should get all the Columns through first stats block. I have changed the function name and comment to reflect that.